### ISSUE TITLE
Typescript slow to show translation keys in intellisense

### ISSUE THREAD

**Issue by Bramgus12** [i18next/i18next#1956]
### Discussed in https://github.com/i18next/i18next/discussions/1955

<div type='discussions-op-text'>

<sup>Originally posted by **Bramgus12** May  5, 2023</sup>
Hi, I have some trouble with using typescript with i18n and react.

When I try to use the `t` function in my code and specifically as a child in tsx, typescript is slow with loading the types that I can use in that function. Can sometimes take up to 5 mins to load. While it is loading typescript itself is so busy getting the types that it won't do anything else. This is really annoying if you have to keep restarting your ts server to make everything work again. My colleagues also tried to do the same thing as I did and had almost no issues (max 3 seconds to load the types). Only difference is that I have a MacBook pro m1-pro while my colleagues are using windows/linux. The translation files that I have a really small (max 30 translations per file).
Anyone know what I am doing wrong?

Here is my setup: 
![image](https://user-images.githubusercontent.com/42220620/236499712-b8748195-eeac-4cc3-9fbe-cd9f12f4efab.png)

i18n.ts
```ts
import i18n from "i18next";
import { initReactI18next } from "react-i18next";

import Backend from "i18next-http-backend";
import LanguageDetector from "i18next-browser-languagedetector";
import componentsENGB from "./languages/en_GB/components.json";
import generalENGB from "./languages/en_GB/general.json";
import componentsNLNL from "./languages/nl_NL/components.json";
import generalNLNL from "./languages/nl_NL/general.json";
import layoutsNLNL from "./languages/nl_NL/layouts.json";
import layoutsENGB from "./languages/en_GB/layouts.json";
import decisionToolNLNL from "./languages/nl_NL/decisionTool.json";
import decisionToolENGB from "./languages/en_GB/decisionTool.json";

export const defaultNS = "general";
export const resources = {
    en_GB: {
        components: componentsENGB,
        general: generalENGB,
        layouts: layoutsENGB,
        decisionTool: decisionToolENGB,
    },
    nl_NL: {
        components: componentsNLNL,
        decisionTool: decisionToolNLNL,
        general: generalNLNL,
        layouts: layoutsNLNL,
    },
} as const;

export const i18nInit = i18n.use(Backend)
    .use(LanguageDetector)
    .use(initReactI18next)
    .init({
        fallbackLng: "en_GB",
        // TODO: debug should only be true on development
        debug: true,
        interpolation: {
            escapeValue: false,
        },
        ns: ["components", "general", "layouts", "decisionTool"],
        defaultNS,
        resources,
    });

export default i18n;
```

i18next.d.ts
```ts
import { defaultNS, resources } from "./i18n";

declare module "i18next" {
    interface CustomTypeOptions {
        defaultNS: typeof defaultNS,
        resources: typeof resources["en_GB"],
    }
}
```

What it looks like when it is loading
![image](https://user-images.githubusercontent.com/42220620/236500163-a27bfe2e-daad-4ac5-af51-643310e280c2.png)
</div>

**Comment by adrai** [comment]
Will probably be addressed with #1911

**Comment by Bramgus12** [comment]
@adrai any reason why on pc it is faster than mac?

**Comment by adrai** [comment]
Please try with i18next v23.0.1 and react-i18next v13.0.0

**Comment by adrai** [comment]
Looks correct...
But if only your IDE does not work, that sounds to be a local problem.
Please try this example in your IDE: https://github.com/i18next/i18next/tree/master/examples/typescript
That works 100% also in IDE, if not, there is something different on your machine.

**Comment by Bramgus12** [comment]
So I did check now with colleagues. It does not work for them as well. It did on an older version, but not on this version.
Also in the example project you have it does not seem to work. Both in vscode and in webstorm.
Below is a screenshot where intellisense does not know what string literals are possible, it just suggests strings that I have typed before.

![image](https://github.com/i18next/i18next/assets/42220620/d298421f-4a40-4a8f-9130-5628289a2173)

It does however seem to know what types are available for that param, but it just doesn't show them inside of the intellisense.

![image](https://github.com/i18next/i18next/assets/42220620/1855d140-2102-46db-b5a8-d410b845ccfd)

**Comment by adrai** [comment]
probably you have some plugin/extension that influences the output?

**Comment by dd-jonas** [comment]
I'm also facing the issue that intellisense doesn't show up anymore using 23.0.2 (VS Code).

**Comment by Bramgus12** [comment]
I tested this on a vscode install without any extensions and it still does not work. Maybe it is the node version that everyone uses? I use version 18.

**Comment by adrai** [comment]
> I tested this on a vscode install without any extensions and it still does not work. Maybe it is the node version that everyone uses? I use version 18.

me too (Node.js v18)

**Comment by dd-jonas** [comment]
Hmm, I cleared my node_modules and did a fresh install and still facing the same issue (also Node 18).

I opened the example in the Github web editor for a quick test.

No intellisense for the t function:
<img width="479" alt="Screenshot 2023-06-16 at 16 47 11" src="https://github.com/i18next/i18next/assets/69415794/1dd2da16-f418-4a84-a349-f69e021c377a">

Intellisense for anything else:
<img width="635" alt="Screenshot 2023-06-16 at 16 47 47" src="https://github.com/i18next/i18next/assets/69415794/c2f8e2dc-ee93-4b6e-bd30-980600afd3c5">

**Comment by adrai** [comment]
Curiosity: I just update visual code (newest version) and now I also don't get the intellisense anymore... wtf?!?
@pedrodurek any idea?

**Comment by Bramgus12** [comment]
I have been looking through the release notes for vscode in may. Don't see anything related to our problem I think. Also have been checking issues to see if anyone else has this issue and also can't seem to find anything :/

**Comment by adrai** [comment]
Probably the update to TS5+ ?

**Comment by adrai** [comment]
https://code.visualstudio.com/updates/v1_79#_typescript-51

**Comment by Bramgus12** [comment]
that would mean that on typescript 4.9 it would be fine and work? Or is it the update of ts within vscode itself?

**Comment by dd-jonas** [comment]
I installed typescript 4.9.3 and pointed vs code to use that version. Doesn't seem to change anything :/

This is the type inference I'm getting. Notice that it's just `string | string[]` as key with the return type `never`.
<img width="743" alt="Screenshot 2023-06-16 at 18 47 50" src="https://github.com/i18next/i18next/assets/69415794/e52ff613-cb14-4a0b-8141-b9eef64d6608">

**Comment by adrai** [comment]
Here the proof it works with the older Visual Code version => https://code.visualstudio.com/updates/v1_78
v1.78.2
![vscode_1 78 2](https://github.com/i18next/i18next/assets/1086194/ff9e7de6-6cc5-4756-8b8c-8b3187c865e4)
v1.79.1
![vscode_1 79 1](https://github.com/i18next/i18next/assets/1086194/489cd3bb-fed3-42e3-ad74-e69cad960b68)

**Comment by adrai** [comment]
Seems the intellisense suggestions appear when Ctrl + Space without inserting anything inside the parenthesis: https://github.com/microsoft/vscode/issues/185381#issuecomment-1595717887

**Comment by pedrodurek** [comment]
It seems to be an issue with typescript 5.1, I opened an issue for it. https://github.com/microsoft/TypeScript/issues/54710

**Comment by Bramgus12** [comment]
From what I can see is this issue being solved in typescript 5.1 somewhere and as it is not an issue with i18next I will close this issue.
